lib/uniris/networking/ip_lookup/ip_lookup.ex

```diff
@@ -0,0 +1,28 @@
+defmodule Uniris.Networking.IPLookup do
+  @moduledoc false
+
+  alias __MODULE__.{Nat, Ipify}
+
+  # Public
+
+  @spec get_node_ip() :: {:ok, :inet.ip_address()} | {:error, :invalid_ip_provider | :not_recognizable_ip | :ip_discovery_error}
+  def get_node_ip do
+    with config when not is_nil(config) <- Application.get_env(:uniris, Uniris.Networking),
```

Comment:
I propose to change a bit the logic. 

Instead of getting external IP and support switch when there is not config, get the provider implementation (default to NAT), if NAT try to fetch it, otherwise switch to IPFY. 
For instance through pattern matching.

So with this scheme, each ip provider could have its fallback resolution.

Reply:
Not sure I understand the difference with the current logic.

Reply:
How I understand the current logic is like:
- If conf is nil -> get external ip
- If conf is present, fetch from the provider

 A proposition could be:
 
 ```elixir
def get_node_ip do
  conf = Application.get_env(:uniris, Uniris.Networking)
  ip_provider = Keyword.get(config, :ip_provider, Nat)
  do_get_node_ip(ip_provider)
end
 
 defp do_get_node_ip(Nat) do
    case Nat.get_node_ip do
       {:ok, ip} ->
          {:ok, ip}
      {:error, _} ->
          IPFY.get_node_ip()
    end
end

defp do_get_node_ip(IPFY), do: IPFY.get_node_ip()
defp do_get_node_ip(Static), do: Static.get_node_ip()
 ```
 
 Where each provider can define, if needed,  a fallback method

Reply:
I propose the next algorithm:
  1. No provider in config -> use NAT discovery.
  2. No provider in config && NAT discovery failed -> use IPIFY.
  3. No provider in config && NAT discovery failed && IPIFY failed (for any reason) -> use hostname from config.
  4. No provider in config && NAT discovery failed && IPIFY failed (for any reason) && no hostname in config -> use localhost.
  5. Provider in config -> use provider.
  6. Provider in config failed -> Goto 1.

Reply:
1) I think it should not really happen as the default will be NAT
2) I think it should be: NAT discovery failed -> use IPFY
3) not sure if the right fallback from IPFY should be an hostname from config. As it's not the expected behavior of auto IP lookup provided by NAT/IPFY implementation
4) I think the localhost should be used only for dev, or private network but this should be the responsibility of a configuration from the Static provider